Media & Text Block: Add vertical alignment options

frontdevde opened the pull request:
## Description

This PR aims to add vertical alignment options to the Media & Text Block.

**Important Note:**

Instead of the initially custom vertical alignment implementation, this PR now builds upon the great work that @getdave has been doing to streamline this functionality by creating the `BlockVerticalAlignmentToolbar` component. Thus this PR will remain in Draft stage until the PR https://github.com/WordPress/gutenberg/pull/13899, which includes the `BlockVerticalAlignmentToolbar` component, has been merged.

## Screenshots 

![media-text-block-valign](https://user-images.githubusercontent.com/1562646/53133338-c1603e80-3540-11e9-8db5-a7d309214a09.gif)

## Todo
- [ ] Merge PR https://github.com/WordPress/gutenberg/pull/13899 to make use of `BlockVerticalAlignmentToolbar`


## How has this been tested?
- [x] Manually verified ability to vertically align content to top/middle/bottom
- [x] Manually verified vertical alignment preview via block editor in `editor` mode
- [x] Manually verified save output on website aligns correctly in various permutations of alignment
- [ ] Manually verified enhancement doesn't require migration path (ie: `deprecation`) when upgrading from old form of the Block to the one in this PR (you can do this by switching to `master` adding the old Block, publishing/updating the post, then switching back to this PR and testing)
- [ ] Manually tested how valign behaves in Responsive (ie: non-desktop) context


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

frontdevde committed:
Add vertical alignment functionality to the media and text block

getdave commented on packages/block-library/src/media-text/editor.scss:
```diff
@@ -2,6 +2,7 @@
 	grid-template-areas:
 		"media-text-media media-text-content"
 		"resizer resizer";
+	align-items: center;
```
Will adding this risk breaking existing Themes that have relied on there being no valign functionality? When we introduce this enhancement the goal is that nothing changes _until_ the user selects one of the v-align options.

This relates to the conversation regarding having a default setting applied on the topbar.

frontdevde replied:
This shouldn't change much as `align-items: center;` on the parent simply replaces the explicit `align-self: center;` on the child. Also this is just in the editor, and brings it in line with the styles that are being used on the Frontend in style.scss where `align-items: center;` on the parent is used.

getdave commented on packages/block-library/src/media-text/style.scss:
```diff
@@ -1,8 +1,6 @@
-}
 
 .wp-block-media-text {
+	display: grid;
```
Can I confirm this is to allow the alignment to work on this item?

frontdevde replied:
It's just moving the property from the duplicate selector above.

getdave reviewed (commented):
Thanks for your work on this.

Looks great and works well. Some minor nit picks but nothing critical.

One thing it might be making it even more explicit in the PR description that feedback for _anything_ other than the Media + Text Block (should be provided on my PR else we'll get in a real mess!).

kjellr commented:
Looks good so far, @frontdevde. 

I do have questions about a default state — if we're center aligning things by default, we should probably have that option selected in the toolbar. 

---

Another observation is that the additional margins that exist on these blocks currently mean that top + bottom aligned blocks will never actually line up with top/bottom of the image next to them. 😕

This is really obvious when the InnerBlock isn't being hovered over. In this case, it just looks like the text is being moved slightly up and down by the buttons, rather than being top/bottom aligned.

![media-text-alignment](https://user-images.githubusercontent.com/1202812/53201244-7ce0ab80-35f1-11e9-8dfe-9799cd8e6b61.gif)

I'm not totally sure what we can do about this, but it's possible we can trim some of those margins a bit in this case. @jasmussen any thoughts on that?

frontdevde commented:
Thanks for the feedback, @kjellr 👍

> I do have questions about a default state — if we're center aligning things by default, we should probably have that option selected in the toolbar.

I've had the same question myself before in the context of @getdave's PR  https://github.com/WordPress/gutenberg/pull/13899 that introduces the vertical alignment toolbar component. I'm quoting the reply I got at the time, in the context of said PR:

> "Essentially we should look to avoid introducing deprecations where possible. By enforcing a vertical alignment to top for Columns we may overide styles provided by existing Themes which won’t expect us to introduce an editor driven CSS alignment. If we default to `undefined` then we retain existing functionality for Columns. The user can then choose to activate an alignment overide as required. This approach also avoids us having to write an explicit Block `deprecation`."

--

> Another observation is that the additional margins that exist on these blocks currently mean that top + bottom aligned blocks will never actually line up with top/bottom of the image next to them.

Yes, noticed that as well. Happy for suggestions!

jasmussen commented:
Thanks for the ping. 

Regarding having the button already pre-selected, I think it's okay for it not to. To me this is not so different from us having a Text align left button, but it not being selected despite most text being left aligned. It is also my understanding that the fewer actual properties we set on blocks, the more a theme can style the defaults.

Regarding the innerblocks additional margin — yeah, this is a challenge. Whenever an innerblocks container is added, additional margin is added as well. This is also one of the challenges we have had to deal with for the columns block, where the singular _column_ blocks add unnecessary margin. We've manually zeroed that out, which was a great headache, and kind of a mess. 

It seems like the solution to that _might_ involve #7694. Or it might involve a slight refactor of how block paddings and margins are painted. That's likely a big one, and one to look at separately. Something in the vein of #8350 but good.

getdave commented:
Thanks for the feedback on this. 

> fewer actual properties we set on blocks, the more a theme can style the defaults.

This is the key point here. We should avoid setting defaults that break existing Blocks backwards compatibility.

> innerblocks additional margin

This feels like it could be something that needs to be addressed to ship this. @frontdevde might you be able to do some investigative research and advise on the amount of work that _might_ be involved here? What's the scope of the problem? Then we can determine whether it can be realisitically be addressed within this PR.

jasmussen commented:
I dove in and took a look. It seems #8350 is definitely the concept to revisit. 

The thing right now is that _every block is born with a top and bottom margin_. That includes inner blocks. Normally content has margins applied on its own. I.e. a paragraph has a top and bottom margin. Those margins even collapse according to obtuse rules. 

But the Columns block should not have a top/bottom margin, neither should the _column_ block that nests inside it. Only the content inside should have its own intrinsic margins. 

As mentioned, the columns block currently zeroes out and overrides this global block margin. Which is hacky. Any fix for this PR would likely involve the same. It seems revisiting #8350 separately is a best bet. I can probably take a look next week. But if anyone else feels adventurous, the margin to refactor out lives here: https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/block-list/style.scss#L60 — I am almost convinced that's the margin that makes the innerblocks have margins they shouldn't have. But be aware this margin affects a lot, so it's not a simple thing to fix.

marekhrabe commented:
The functionality looks really solid to me. Extra margins now noticeable with the top/bottom alignments.

I see some extra files in here (like icons.js) but I assume it's just a matter of rebasing again on top of the #13899, right?

frontdevde commented:
> I see some extra files in here (like icons.js) but I assume it's just a matter of rebasing again on top of the #13899, right?

That is correct! 👍

jorgefilipecosta commented on packages/block-library/src/media-text/index.js:
```diff
@@ -66,6 +66,9 @@ const blockAttributes = {
 		default: false,
 	},
+	verticalAlignment: {
+		type: 'string',
```
Would it make sense to have a default value of center?

getdave replied:
We established a convention in the [parent PR](https://github.com/WordPress/gutenberg/pull/13899) (Columns) that defaults would not be applied as this could be confusing for existing Themes. So default to `null` ("leave as is") so as to retain backwards compat.

jorgefilipecosta commented on packages/block-library/src/media-text/editor.scss:
```diff
@@ -10,13 +11,24 @@
 		"resizer resizer";
 }
 
+.wp-block-media-text.is-vertically-aligned-top {
```
It seems these styles duplicate what we have in styles.scss, I think they can be removed.

jorgefilipecosta commented on packages/block-library/src/media-text/index.js:
```diff
@@ -66,6 +66,9 @@ const blockAttributes = {
 		type: 'boolean',
```
It would be nice to add another block fixture in packages/e2e-tests/fixtures/blocks that test a media-text block with a left or right alignment.

frontdevde replied:
Recently noticed this comment when scrolling back through the comment history. Added a fixture that tests for the vertical alignment.

jorgefilipecosta reviewed (commented):
Hi @frontdevde thank you for this enhancements 👍 I focused my review on media & text changes as the other artifacts are already being reviewed in https://github.com/WordPress/gutenberg/pull/13899.
Things worked well in my tests, I just left some suggestions.

m-e-h commented:
This is nice! 
I actually had a request from a user the other day, to make the media-image stretch to match the full-height of the media text. Is that another button that would make sense or should be added here? :arrow_up_down:

frontdevde commented:
> This pull request is still a work in progress.

> Thus this PR will remain in Draft stage until the PR #13899, which includes the BlockVerticalAlignmentToolbar component, has been merged.

I've updated the state of this Draft PR to only show the diff to PR #13899. We'll get back to this once PR #13899 has landed. For now I've added the `Status [Blocked]` label.

getdave commented:
@frontdevde The VAlign PR is now merged. Note that `BlockVerticalAlignmentToolbar` has moved to the `@wordpress/block-editor` package.

https://github.com/WordPress/gutenberg/pull/13899

frontdevde committed:
Remove duplicate styles

frontdevde commented:
@m-e-h Is the feature request you were referring to covered by PR  https://github.com/WordPress/gutenberg/pull/14445 ?

m-e-h commented:
That would be it @frontdevde! Thanks!

getdave commented:
@frontdevde Looks like the automated tests are not working. You may need to ping someone in Core Editor to get these restarted as they moved them to a new platform which may be causing this issue.

getdave commented:
@frontdevde Are we still Blocked here?

mtias commented:
I'd love to have an option here that makes the media cover the full available space, to the left or right (like background cover or content-fit) cropping the image as necessary, etc. I think that was present at some point in this block cc @jorgefilipecosta

m-e-h commented:
@mtias to quote @frontdevde here https://github.com/WordPress/gutenberg/pull/13989#issuecomment-475114718
> Is the feature request you were referring to covered by PR #14445 ?

:smile:

frontdevde committed:
Add fixture for vertical alignment test

frontdevde commented:
> @frontdevde Are we still Blocked here?

No! I forgot to remove the tag, thanks for pointing that out.

Just added a fixture for a vertical alignment test. That resolves the last open comment on this one.

jorgefilipecosta reviewed (approved):
The code changes look good and everything worked as expected in my tests. Thank you for this improvement 👍

frontdevde committed:
Add entry to changelog

frontdevde committed:
Merge branch 'master' into update/media-text-block-v-align

cdnlin commented:
A question from a new-ish Gutenberg user: this is exactly what I need. Is it available to us regular folks yet? Or do you have a date for its availability yet? We sure need it for the new site we're creating!

Thanks,

Linda

noisysocks commented:
Hi @cdnlin, this feature is currently available for sites using the Gutenberg plugin and will soon be available for all sites running WordPress 5.3.

cdnlin commented:
Hey, great news, noisysocks! Thanks so much for letting me know. 5.3 has to be getting close since we're at 5.2.1 now. Until then, appreciate the heads up about the plugin. I haven't downloaded it yet, but will now.
